hazelcast/config.py

```diff
@@ -581,820 +332,1232 @@ class _Config:
-        else:
+    def cluster_members(self, value: typing.List[str]) -> None:
+        if not isinstance(value, list):
             raise TypeError("cluster_members must be a list")
 
-    @property
-    def cluster_name(self):
+        for address in value:
+            if not isinstance(address, str):
+                raise TypeError("cluster_members must be list of strings")
+
+        self._cluster_members = value
+
+    @property
+    def cluster_name(self) -> str:
+        """Name of the cluster to connect to.
+
+        The name is sent as part of the client authentication message and may
+        be verified on the member. By default, set to ``dev``.
+        """
         return self._cluster_name
 
     @cluster_name.setter
-    def cluster_name(self, value):
-        if isinstance(value, str):
-            self._cluster_name = value
-        else:
+    def cluster_name(self, value: str) -> None:
+        if not isinstance(value, str):
             raise TypeError("cluster_name must be a string")
 
+        self._cluster_name = value
+
     @property
-    def client_name(self):
+    def client_name(self) -> typing.Optional[str]:
+        """Name of the client instance.
+
+        By default, set to ``hz.client_${CLIENT_ID}``, where ``CLIENT_ID``
+        starts from ``0`` and it is incremented by ``1`` for each new client.
+        """
         return self._client_name
 
     @client_name.setter
-    def client_name(self, value):
-        if isinstance(value, str):
-            self._client_name = value
-        else:
+    def client_name(self, value: str) -> None:
+        if not isinstance(value, str):
             raise TypeError("client_name must be a string")
 
+        self._client_name = value
+
     @property
-    def connection_timeout(self):
+    def connection_timeout(self) -> float:
+        """Socket timeout value in seconds for the client to connect member
+        nodes.
+
+        Setting this to ``0`` makes the connection blocking. By default, set
+        to ``5.0``.
+        """
         return self._connection_timeout
 
     @connection_timeout.setter
-    def connection_timeout(self, value):
-        if isinstance(value, number_types):
-            if value < 0:
-                raise ValueError("connection_timeout must be non-negative")
-            self._connection_timeout = value
-        else:
+    def connection_timeout(self, value: float):
+        if not isinstance(value, number_types):
             raise TypeError("connection_timeout must be a number")
 
+        if value < 0:
+            raise ValueError("connection_timeout must be non-negative")
+
+        self._connection_timeout = value
+
     @property
-    def socket_options(self):
+    def socket_options(self) -> typing.List[typing.Tuple[int, int, typing.Union[int, bytes]]]:
+        """List of socket option tuples.
+
+        The tuples must contain the parameters passed into the
+        :func:`socket.setsockopt` in the same order.
```

Comment:
But, we are depending on it to set socket options. I don't see a problem in depending on it for documentation as well

Reply:
We don't depend on the socket options in our public API, so in the future we can modify these values after getting these parameters or use a totally different thing without breaking user's code. In order to be completely independent, I think it would be nice to make the docs independent too. In any case, it's not easy to see what are options for `socket.setsockopt` unless I check out the main Python documentation.

Reply:
We are depending on it by saying that we will pass whatever socket options you passed to us directly to it. We are not independent of that. So, the options you have there are available here as well. I don't like the idea of duplicating some parts of the standard library documentation. If somebody wants to use this, they can go to the standard library docs for their python version and check it from there. There are lots of options, some are platform specific, some are added in some later Python versions, and some are even not documented at all but they give references to man pages. We can't sync with that, so I believe our best option is just to give a reference like here.